### ISSUE TITLE
[Bug]: AI/GS settings randomization issues

### ISSUE THREAD

**Issue by SamuXarick** [OpenTTD/OpenTTD#10341]
### Version of OpenTTD

13.0-RC1, Windows 11 Pro

### Expected result

N/A

### Actual result

N/A

### Steps to reproduce

There's several issues, but they're all related to each other so I'm submiting them together. Summary:
Bug 1 - `start_date` setting in a Random AI configuration does not have its random deviation applied upon New Game start, ...
Bug 2 - ... but once the Random AI starts its company, random deviation is applied to `start_date` which by this time, it's pointless, the company has already started.
Bug 3 - AI settings that have the flag `SCRIPT_RANDOM` are rerandomized upon savegame load.

Now, describing the rules for which flavour random is intended to happen and when is also kind of complex. But let me try:
A - Random AIs that become a real AI once their companies start, are intended to have all their `SCRIPT_RANDOM` flagged settings to have the values randomized. Non-Random AIs however, don't follow this rule. This is intended, and by design.
B - `random_deviation` settings for either Random and Non-Random AIs, are intended to have their values deviated, but only once, upon New Game start for Non-Random AIs, or upon company start for Random AIs.

How the Loading of savegames should work, given those rules above:
- `SCRIPT_RANDOM` flagged settings should not occur for AIs that have already started, no matter how they were started, Random AI or not. For AIs that have yet to start, they should follow the rules in point A above.
- `random_deviation` should not occur, except for Random AIs that have not yet started, but only for settings other than `start_date`.

AI script used for testing: [AI Settings test.zip](https://github.com/OpenTTD/OpenTTD/files/10406107/AI.Settings.test.zip)
This test AI contains 6 settings, each with its specific perks, like being a `random_deviation` setting or being a `SCRIPT_RANDOM` or `SCRIPT_INGAME`, to help reproduce the issue. The non `SCRIPT_BOOLEAN` settings have in common a minimum value of 0, a maximum value of 200, a default value of 100, and the `random_deviation` settings have a deviation value of 50.

Now for the reproducing steps. I took various screenshots depicting 4 AIs and the values of their settings over different stages of the game.
First stage is the 'before New Game' stage, where you setup the AIs.
Second stage is the 'after New Game' stage, where the AIs have not yet started.
Third stage is the 'after the first 2 AIs have started' stage, where you also create a savegame.
Fourth stage is the 'after loading the savegame' stage.
Make sure this is the only AI installed, so that when it starts as Random AI, it is it.

Stage 1 - Setup 4 AIs, with all settings set to default, as shown in the screenshots.
![AI1 Main Menu](https://user-images.githubusercontent.com/43006711/212184927-d85f9c77-a572-48d7-a0bd-a5710a929fac.png)
![AI2 Main Menu](https://user-images.githubusercontent.com/43006711/212184943-14640f1a-ef31-449d-ba93-620e1a615c25.png)
![AI3 Main Menu](https://user-images.githubusercontent.com/43006711/212184965-fdeff430-c78d-4aea-a20b-b6e337ddbf0b.png)
![AI4 Main Menu](https://user-images.githubusercontent.com/43006711/212184986-827bee9c-dcbe-49d2-9577-af4802674c7d.png)

Stage 2 - When ready, start a new game, and check their settings again
![AI1 New Game](https://user-images.githubusercontent.com/43006711/212198575-622177ec-df63-4c4f-b0e7-764def402cbc.png)
Bug 1 - `start_date` wasn't deviated.

![AI2 New Game](https://user-images.githubusercontent.com/43006711/212198612-cc3dc8f2-7dd1-4520-bd00-5a90731efa4d.png)
No issues here. Random deviation was applied to `start_date` and the `random_deviation` settings. As this is a Non-Random AI, it is also intended that the `SCRIPT_RANDOM` settings are not randomized.

![AI3 New Game, Saved as](https://user-images.githubusercontent.com/43006711/212198627-d53caac0-8622-409d-b2d5-8216d323e669.png)
Bug 1 - `start_date` wasn't deviated.

![AI4 New Game, Saved as](https://user-images.githubusercontent.com/43006711/212198649-66a98ddb-18fe-4127-b41a-12ee570394bc.png)
No issues here. Random deviation was applied to `start_date` and the `random_deviation` settings. As this is a Non-Random AI, it is also intended that the `SCRIPT_RANDOM` settings are not randomized.

Stage 3 - Let the first 2 AIs start, then check their settings again and create a savegame
![AI1 Started, Saved as](https://user-images.githubusercontent.com/43006711/212198786-5a2c5f0c-6808-4199-b2a6-65e900400b00.png)
Bug 2 - `start_date` value was deviated too late. But the rest is fine. As this was initially setup as a Random AI, it is intended to have `random_deviation` and `SCRIPT_RANDOM` settings to randomize. the `SCRIPT_BOOLEAN` one is difficult to discern, but I can assure you it just randomized to true.

![AI2 Started, Saved as](https://user-images.githubusercontent.com/43006711/212198800-b4fc03fd-daf6-4317-b2b9-c8141f6f4dce.png)
No issues here. As per the rules, the already deviated values remained the same and the `SCRIPT_RANDOM` settings were not randomized as intended.

Stage 4 - Load the savegame and check their settings one last time
![AI1 Started, Loaded as](https://user-images.githubusercontent.com/43006711/212198965-3ce183b1-dc59-437a-8655-82db7af5d8bd.png)
Surprisingly, no issues here. All the values that have been saved, are loaded back correctly.

![AI2 Started, Loaded as](https://user-images.githubusercontent.com/43006711/212198982-f9f05f4c-b86e-43c6-b9fb-46b43b532ebd.png)
Bug 3 - `SCRIPT_RANDOM` settings have been rerandomized, but notice that it were only those that are still able to be modified in-game, with the `SCRIPT_INGAME` flag.

![AI3 Loaded as](https://user-images.githubusercontent.com/43006711/212199001-10a481f7-4e44-4c16-af60-49c75747891c.png)
No issue here, even though the `start_date` wasn't deviated previously in stage 2.

![AI4 Loaded as](https://user-images.githubusercontent.com/43006711/212199010-d0e951ab-c370-4b17-bdf9-b72ecd1a27b8.png)
Bug 3 - `SCRIPT_RANDOM` settings have been rerandomized. Because the AI hasn't started yet, the `SCRIPT_INGAME` flag takes no effect here.

**Comment by glx22** [comment]
`start_date` issues should go away with #10330.
There's probably a way to figure out if the AI is starting in `AfterLoadGame()` or not.

**Comment by SamuXarick** [comment]
It's so good that `start_date` is out of the way! Bug 1 and 2 are gone! Helps in isolating bug 3.

For bug 3, I investigated this one more time, now on 20240128-master-g05b253b33b.

> When starting a new game:
> 1 - Adds random deviation to parameter5 and parameter6 for yet to start AIs
> ->	openttd.exe!ScriptConfig::AddRandomDeviation() Line 147	C++
>  	openttd.exe!ScriptConfig::ScriptConfig(const ScriptConfig * config) Line 62	C++
>  	openttd.exe!AIConfig::AIConfig(const AIConfig * config) Line 29	C++
>  	openttd.exe!MakeNewgameSettingsLive() Line 364	C++
> 
> When starting an AI company after new game or after loading a savegame:
> 2 - Randomizes SCRIPTCONFIG_RANDOM parameter1, parameter2, parameter3, parameter4 for Random AIs
>   - Adds random deviation to parameter5 and parameter6 for Random AIs
>   - Nothing for yet to start AIs
> ->	openttd.exe!ScriptConfig::AddRandomDeviation() Line 147	C++
> -> 	openttd.exe!ScriptConfig::Change(std::optional<std::string const> name, int version, bool force_exact_match, bool is_random) Line 45	C++
>  	openttd.exe!AI::StartNew(Owner company, bool rerandomise_ai) Line 51	C++
>  	openttd.exe!DoStartupNewCompany(bool is_ai, Owner company) Line 592	C++
> 
> When loading a savegame:
> 1 - Randomizes SCRIPTCONFIG_RANDOM parameter1, parameter2, parameter3, parameter4 for started AIs, Random AIs and yet to start AIs
>   - Adds random deviation to parameter5 and parameter6 for started AIs, Random AIs and yet to start AIs
> ->	openttd.exe!ScriptConfig::AddRandomDeviation() Line 147	C++
> ->	openttd.exe!ScriptConfig::Change(std::optional<std::string const> name, int version, bool force_exact_match, bool is_random) Line 41	C++
>  	openttd.exe!AIPLChunkHandler::Load() Line 91	C++
>  	openttd.exe!SlLoadChunk(const ChunkHandler & ch) Line 1973	C++
>  	openttd.exe!SlLoadChunks() Line 2128	C++
>  	openttd.exe!DoLoad(LoadFilter * reader, bool load_check) Line 2999	C++
> 
> 2 - Sets values from the saved data to parameter1, parameter3, parameter5 and parameter6 for started AIs
>   - Sets values from the saved data to parameter1, parameter2, parameter3, parameter4, parameter5, parameter6 for started Random AIs
>   - Sets values from the saved data to parameter5 and parameter6 for yet to start AIs
> ->	openttd.exe!ScriptConfig::StringToSettings(const std::string & value) Line 187	C++
>  	openttd.exe!AIPLChunkHandler::Load() Line 116	C++
>  	openttd.exe!SlLoadChunk(const ChunkHandler & ch) Line 1973	C++
>  	openttd.exe!SlLoadChunks() Line 2128	C++
>  	openttd.exe!DoLoad(LoadFilter * reader, bool load_check) Line 2999	C++


|                                                   | parameter1 RANDOM | parameter2 RANDOM INGAME | parameter3 BOOLEAN RANDOM | parameter4 BOOLEAN RANDOM INGAME | parameter5 deviation | parameter6 deviation INGAME |
| :------------------------------------------------ | :--------: | :--------: | :--------: | :--------: | :--------: | :--------: |
| When loading a savegame (started AI)              | randomized, restored :green_circle: | randomized, missing :red_circle: | randomized, restored :green_circle: | randomized, missing :red_circle: | deviated, restored :green_circle:  | deviated, restored :green_circle:  |
| When loading a savegame (started Random AI)       | randomized, restored :green_circle: | randomized, restored :green_circle: | randomized, restored :green_circle: | randomized, restored :green_circle: | deviated, restored :green_circle:  | deviated, restored :green_circle:  |
| When loading a savegame (yet to start AI)         | randomized, missing :red_circle: | randomized, missing :red_circle: | randomized, missing :red_circle: | randomized, missing :red_circle: | deviated, restored :green_circle:  | deviated, restored :green_circle:  |

**Comment by glx22** [comment]
I don't fully understand your description, but if a setting is not restored from the save then it was using the default value and never have been randomized before saving.

**Comment by TrueBrain** [comment]
This ticket is just a personal brain-dump. I am sure it has value to you, but for us, it is just a bunch of text without structure. From what I can deduce:

- `start_date` issue is resolved
- Rest is duplicate of https://github.com/OpenTTD/OpenTTD/issues/9540

Going to close this issue.

**Comment by SamuXarick** [comment]
> I don't fully understand your description, but if a setting is not restored from the save then it was using the default value and never have been randomized before saving.

The load is doing the (re)randomization of a setting which was a default value (hence it was not saved).
